File: lib/std/core/mapbounds.lua
require 'std.core.types.region'
require 'std.core.types.rect'
local hook = require 'std.core.hook'

mapbounds = mapbounds or {}

function mapbounds.playableRect()
    return mapbounds.__playable
end

function mapbounds.playableRegion()
    return mapbounds.__playableRegion
end

function mapbounds.wholeRect()
    return mapbounds.__whole
end

function mapbounds.wholeRegion()
    return mapbounds.__wholeRegion
end

hook.add("init", function()
    mapbounds.__playable = rectFromNative(GetPlayableMapRect())
    mapbounds.__whole = rectFromNative(GetWorldBounds())

    local playableRegion = region.new()
    playableRegion:addRect(mapbounds.__playable)
    mapbounds.__playableRegion = playableRegion

    local wholeRegion = region.new()
    wholeRegion:addRect(mapbounds.__whole)
    mapbounds.__wholeRegion = wholeRegion
end)

File: ts/src/tsstd/core/mapbounds.ts
import Rect from "./types/rect"
import Region from "./types/region"

let context = util.contextFn("mapbounds", () => {
    let playable = Rect.of(GetPlayableMapRect())
    let bounds = Rect.of(GetWorldBounds())

    let playableRegion = Region.create()
    let boundsRegion = Region.create()
    playableRegion.addRect(playable)
    boundsRegion.addRect(bounds)

    return {
        playable: playable,
        bounds: bounds,
        playableRegion: playableRegion,
        boundsRegion: boundsRegion
    }
})

export function playableRect(): Rect {
    return context.playable
}

export function playableRegion(): Region {
    return context.playableRegion
}

export function wholeRect(): Rect {
    return context.bounds
}

export function wholeRegion(): Region {
    return context.boundsRegion
}
